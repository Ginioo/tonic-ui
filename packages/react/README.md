# @tonic-ui/react

A comprehensive React component library for Tonic UI.

## Installation

```sh
// with npm
npm install @tonic-ui/react

// with yarn
yarn add @tonic-ui/react
```
