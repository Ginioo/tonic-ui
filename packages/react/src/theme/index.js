import ThemeProvider from './ThemeProvider';
import createTheme from './createTheme';
import useTheme from './useTheme';
import theme from './theme';

export {
  ThemeProvider,
  createTheme,
  theme,
  useTheme,
};
