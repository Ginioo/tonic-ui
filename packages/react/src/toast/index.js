import Toast from './Toast';
import ToastCloseButton from './ToastCloseButton';
import ToastContainer from './ToastContainer';
import ToastController from './ToastController';
import ToastIcon from './ToastIcon';
import ToastManager from './ToastManager';
import ToastMessage from './ToastMessage';
import ToastTransition from './ToastTransition';
import ToastTransitionGroup from './ToastTransitionGroup';
import useToastManager from './useToastManager';

export {
  Toast,
  ToastCloseButton,
  ToastContainer,
  ToastController,
  ToastIcon,
  ToastManager,
  ToastMessage,
  ToastTransition,
  ToastTransitionGroup,
  useToastManager,
};
