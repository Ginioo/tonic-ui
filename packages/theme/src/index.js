import createTheme from './createTheme'; // deprecated

const theme = createTheme('rem');

export { createTheme };
export default theme;
