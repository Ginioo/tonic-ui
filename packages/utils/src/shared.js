import { ensureArray, ensureBoolean, ensureString } from 'ensure-type';
import { isPlainObject } from './assertion';

const _joinWords = (words) => {
  words = ensureArray(words);
  if (words.length === 0) {
    return '';
  }
  if (words.length === 1) {
    return `'${words[0]}'`;
  }
  if (words.length === 2) {
    return `'${words[0]}' and '${words[1]}'`;
  }
  return `'${words.slice(0, -1).join('\', \'')}', and '${words.slice(-1)}'`;
};

const _deepClone = (source, seen = new WeakMap()) => {
  // Use a `WeakMap` to track objects and detect circular references.
  // If the object has been cloned before, return the cached cloned version.
  if (seen.has(source)) {
    return seen.get(source);
  }

  if (Array.isArray(source)) {
    const clonedArray = [];
    seen.set(source, clonedArray);
    for (let i = 0; i < source.length; ++i) {
      clonedArray[i] = _deepClone(source[i], seen);
    }
    return clonedArray;
  }

  if (isPlainObject(source)) {
    const clonedObject = {};
    seen.set(source, clonedObject);
    for (const [key, value] of Object.entries(source)) {
      clonedObject[key] = _deepClone(value, seen);
    }
    return clonedObject;
  }

  // For primitive values and other types, return as is
  return source;
};

export const ariaAttr = (condition) => {
  return ensureBoolean(condition) ? true : undefined;
};

export const callAll = (...fns) => {
  return function mergedFn(...args) {
    fns.forEach((fn) => {
      fn?.(...args);
    });
  };
};

export const callEventHandlers = (...fns) => {
  return function mergedFn(event) {
    fns.some((fn) => {
      fn?.(event);
      return event?.defaultPrevented;
    });
  };
};

export const dataAttr = (condition) => {
  return condition ? '' : undefined;
};

export const merge = (target, source, options = { clone: true }) => {
  // Merge arrays
  if (Array.isArray(target) && Array.isArray(source)) {
    const output = options.clone ? [...target] : target;
    source.forEach((item, index) => {
      if (isPlainObject(item) && isPlainObject(output[index])) {
        output[index] = merge(output[index], item, options);
      } else {
        output[index] = options.clone ? _deepClone(item) : item;
      }
    });
    return output;
  }

  // Merge plain objects
  if (isPlainObject(target) && isPlainObject(source)) {
    const output = options.clone ? { ...target } : target;
    for (const [key, value] of Object.entries(source)) {
      if (isPlainObject(value) && Object.prototype.hasOwnProperty.call(output, key) && isPlainObject(output[key])) {
        output[key] = merge(output[key], value, options);
      } else {
        output[key] = options.clone ? _deepClone(value) : value;
      }
    }
    return output;
  }

  return options.clone ? _deepClone(source) : source;
};

export const noop = () => {};

export const once = (fn) => {
  let called = false;
  let result;
  return (...args) => {
    if (!called) {
      result = fn(...args);
      called = true;
    }
    return result;
  };
};

export const runIfFn = (valueOrFn, ...args) => {
  return (typeof valueOrFn === 'function') ? valueOrFn(...args) : valueOrFn;
};

export const warnDeprecatedProps = (props, options) => {
  const prefix = options?.prefix ?? 'Warning:';
  const alternative = ensureArray(options?.alternative);
  const willRemove = ensureBoolean(options?.willRemove);
  const message = ensureString(options?.message);

  props = ensureArray(props);
  if (props.length === 0) {
    return;
  }

  const messages = [prefix];
  const verb = (props.length > 1) ? 'are' : 'is';

  if (willRemove) {
    messages.push(`${_joinWords(props)} ${verb} deprecated and will be removed in the next major release.`);
  } else {
    messages.push(`${_joinWords(props)} ${verb} deprecated.`);
  }

  if (alternative.length > 0) {
    messages.push(`Use ${_joinWords(alternative)} instead.`);
  }

  if (message) {
    messages.push(message);
  }

  console.error(messages.join(' '));
};

export const warnRemovedProps = (props, options) => {
  const prefix = options?.prefix ?? 'Warning:';
  const alternative = ensureArray(options?.alternative);
  const message = ensureString(options?.message);

  props = ensureArray(props);
  if (props.length === 0) {
    return;
  }

  const messages = [prefix];
  const verb = (props.length > 1) ? 'are' : 'is';

  messages.push(`${_joinWords(props)} ${verb} removed.`);

  if (alternative.length > 0) {
    messages.push(`Use ${_joinWords(alternative)} instead.`);
  }

  if (message) {
    messages.push(message);
  }

  console.error(messages.join(' '));
};
