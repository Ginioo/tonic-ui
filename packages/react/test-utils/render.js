import { matchers } from '@emotion/jest';
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import * as React from 'react';
import {
  PortalManager,
  ToastManager,
  TonicProvider,
  createTheme,
} from '../src';

// https://emotion.sh/docs/@emotion/jest#tohavestylerule
//
// Add the custom matchers provided by '@emotion/jest'
expect.extend(matchers);

const customRender = (ui, options) => {
  const {
    theme: themeOptions = {
      cssVariables: {
        prefix: 'tonic',
        rootSelector: ':root',
      },
    },
    ...rest
  } = { ...options };

  const wrapper = ({ children }) => (
    <TonicProvider
      colorMode={{
        defaultValue: 'dark',
      }}
      theme={createTheme(themeOptions)}
    >
      <PortalManager>
        <ToastManager>
          {children}
        </ToastManager>
      </PortalManager>
    </TonicProvider>
  );

  return render(ui, { wrapper, ...rest });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
