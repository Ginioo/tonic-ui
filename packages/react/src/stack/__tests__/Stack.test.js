import { screen } from '@testing-library/react';
import { testA11y } from '@tonic-ui/react/test-utils/accessibility';
import { render } from '@tonic-ui/react/test-utils/render';
import { Box, Stack } from '@tonic-ui/react/src';
import React from 'react';

describe('Stack', () => {
  it('should render correctly', async () => {
    const renderOptions = {
      useCSSVariables: true,
    };
    const { container } = render((
      <Stack spacing="4x">
        <Box />
        <Box />
        <Box />
      </Stack>
    ), renderOptions);

    expect(container).toMatchSnapshot();

    await testA11y(container);
  });

  it('should apply the default direction of column', () => {
    render(
      <Stack data-testid="stack">
        <Box />
        <Box />
      </Stack>
    );

    const stack = screen.getByTestId('stack');
    expect(stack).toHaveStyle('flex-direction: column');
  });

  it('should apply the custom direction when specified', () => {
    render(
      <>
        <Stack data-testid="stack-row" direction="row">
          <Box />
          <Box />
        </Stack>
        <Stack data-testid="stack-column" direction="column">
          <Box />
          <Box />
        </Stack>
      </>
    );

    expect(screen.getByTestId('stack-row')).toHaveStyle('flex-direction: row');
    expect(screen.getByTestId('stack-column')).toHaveStyle('flex-direction: column');
  });

  it('should apply spacing between child elements', () => {
    const renderOptions = {
      useCSSVariables: true,
    };
    render((
      <Stack data-testid="stack" spacing="2x">
        <Box />
        <Box />
      </Stack>
    ), renderOptions);

    const stack = screen.getByTestId('stack');
    expect(stack).toHaveStyle('row-gap: var(--tonic-sizes-2x)');
  });

  it('should not wrap each child by default when shouldWrapChildren is false', () => {
    render(
      <Stack data-testid="stack">
        <Box data-testid="stack-item-1" />
        <Box data-testid="stack-item-2" />
      </Stack>
    );

    const stack = screen.getByTestId('stack');
    expect(stack.children).toHaveLength(2);
    expect(stack.children[0]).toHaveAttribute('data-testid', 'stack-item-1');
    expect(stack.children[1]).toHaveAttribute('data-testid', 'stack-item-2');
  });

  it('should wrap each child when shouldWrapChildren is true', () => {
    render(
      <Stack data-testid="stack" shouldWrapChildren>
        <Box data-testid="stack-item-1" />
        <Box data-testid="stack-item-2" />
      </Stack>
    );

    const stack = screen.getByTestId('stack');
    expect(stack.children).toHaveLength(2);
    expect(stack.children[0].firstChild).toHaveAttribute('data-testid', 'stack-item-1');
    expect(stack.children[1].firstChild).toHaveAttribute('data-testid', 'stack-item-2');
  });
});
