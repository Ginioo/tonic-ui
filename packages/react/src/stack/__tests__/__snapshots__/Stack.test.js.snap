// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack should render correctly 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  row-gap: var(--tonic-sizes-4x);
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-2 emotion-1"
    />
    <div
      class="emotion-2 emotion-1"
    />
    <div
      class="emotion-2 emotion-1"
    />
  </div>
</div>
`;
